src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php

```diff
@@ -76,11 +91,186 @@ public function indexAction(Request $request, MerchandiseReturnFilters $filters)
+
+        return $this->redirectToRoute(
+                'admin_merchandise_returns_edit',
+                [
+                    'merchandiseReturnId' => $merchandiseReturnId,
+                ]
+        );
+    }
+
+    /**
+     * Deletes cartRules on bulk action
+     *
+     * @DemoRestricted(redirectRoute="admin_merchandise_returns_index")
+     *
+     * @param Request $request
+     *
+     * @return RedirectResponse
+     */
+    public function bulkDeleteProductAction(int $merchandiseReturnId, Request $request): RedirectResponse
+    {
+        $merchandiseReturnProducts = $this->getBulkMerchandiseReturnProductsFromRequest($request);
+
+        try {
+            $this->getCommandBus()->handle(
+                new BulkDeleteProductFromMerchandiseReturnCommand(
+                    $merchandiseReturnId,
+                    $merchandiseReturnProducts
+                )
+            );
+            $this->addFlash(
+                'success',
+                $this->trans('Successful deletion.', 'Admin.Notifications.Success')
+            );
+        } catch (Exception $e) {
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages($e)));
+        }
+
+        return $this->redirectToRoute('admin_merchandise_returns_index');
+    }
+
+    /**
+     * Provides cart rule ids from request of bulk action
+     *
+     * @param Request $request
+     *
+     * @return MerchandiseReturnProduct[]
+     */
+    private function getBulkMerchandiseReturnProductsFromRequest(Request $request): array
+    {
+        $merchandiseReturnDetailIds = $request->request->get('merchandise_return_products_merchandise_return_bulk');
+        $merchandiseReturnCustomizationIds = $request->request->get('merchandise_return_products_merchandise_return_bulk_id_customization');
+        if (!is_array($merchandiseReturnDetailIds)) {
+            return [];
+        }
+
+        $merchandiseReturnProducts = [];
+
+        foreach ($merchandiseReturnDetailIds as $key => $merchandiseReturnDetailId) {
+            $merchandiseReturnProduct = new MerchandiseReturnProduct(
+                (int) $merchandiseReturnDetailId
+            );
+
+            if ($merchandiseReturnCustomizationIds[$key]) {
+                $merchandiseReturnProduct->setCustomizationId((int) $merchandiseReturnCustomizationIds[$key]);
+            }
+
+            $merchandiseReturnProducts[] = $merchandiseReturnProduct;
+        }
+
+        return $merchandiseReturnProducts;
+    }
+
     /**
      * @return FormHandlerInterface
      */
     private function getOptionsFormHandler(): FormHandlerInterface
     {
         return $this->get('prestashop.admin.merchandise_return_options.form_handler');
     }
+
+    /**
+     * Provides error messages for exceptions
+     *
+     * @return array
+     */
+    private function getErrorMessages(Exception $e = null): array
+    {
+        return [
+            MerchandiseReturnConstraintException::class => [
```

Comment:
Seems like there are a lot of missing error messages compared to the number of exceptions you created

Reply:
Added missing messages